### PR TITLE
Ports paradise's modular buildmode

### PR DESCRIPTION
**Pull request by marlyn-x86**
:cl: Crazylemon
admin: "Fill" Buildmode, "Area Edit" Buildmode, "Boom" Buildmode
refactor: Buildmode is now in modules instead of defining new modes by extra branches of switch statements.
imageadd: Switches buildmode button icons to something with a slightly more refined background.
admin: Buildmode now switches between modes using a discrete switcher, instead of by repeatedly clicking.
/:cl:
Ports ParadiseSS13/Paradise#9291

Features it includes:
+ Datumized buildmode sub-modes, to make the system more extensible.
+ A more user-friendly mode and dir-switching system.
+ The following buildmodes:
    + "Fill" Buildmode
    + "Area Edit" Buildmode
    + "Boom" Buildmode

Things I did not port due to wanting to get the simple stuff through first:
+ ["Atmos" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/atmos.dm)
+ ["Map Saving" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/save.dm)
+ ["Door Link" Buildmode](https://github.com/ParadiseSS13/Paradise/blob/8edd8676be47a1f26b57caf508f108cb301da800/code/modules/buildmode/submodes/link.dm)

Pictures of the new mode switching system:

tgstation_bm_modeswitch

tgstation_bm_dirswitch

### CODE REVIEW

File: code/modules/buildmode/submodes/variable_edit.dm [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/datum/buildmode_mode/varedit/handle_click(user, params, obj/object)
+	var/list/pa = params2list(params)
+	var/left_click = pa.Find("left")
+	var/right_click = pa.Find("right")
+
+	if(left_click)
+		if(object.vars.Find(varholder))
+			log_admin("Build Mode: [key_name(user)] modified [object.name]'s [varholder] to [valueholder]")
+			object.vars[varholder] = valueholder
```
Comment (AnturK) [inline code/modules/buildmode/submodes/variable_edit.dm]: Drops varedit wrappers which is big no no.

**Review by AnturK — CHANGES_REQUESTED** [review]
If selection a region is shared between modes it should be in root datum not repeated everywhere

File: code/modules/buildmode/submodes/variable_edit.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	varholder = null
+	valueholder = null
+
+/datum/buildmode_mode/varedit/change_settings(mob/user)
+	varholder = input(user,"Enter variable name:" ,"Name", "name")
+	
+	if(!vv_varname_lockcheck(varholder))
+		return
+
+	var/temp_value = user.client.vv_get_value()
```
Comment (Crazylemon64) [inline code/modules/buildmode/submodes/variable_edit.dm]: Is this a suitable interfacing with the existing VV system? Without an existing datum to interface with, this is the closest I could get. I made sure to split out the varname locking logic into its own proc, so I could use it here without accidentally allowing unauthorized access if things got changed later.

Also, it looks like the edit var COMSIG doesn't get fired in `object.vv_edit_var`. Should I fire that signal manually in here? Doing so feels like a bad idea.

Comment (ninjanomnom) [reply code/modules/buildmode/submodes/variable_edit.dm]: That signal is meant to trigger once per admin var edit. We only use it for a jokey plush atm which complains when var edits happen.

In other words no don't do anything with it.

Comment (AnturK) [reply code/modules/buildmode/submodes/variable_edit.dm]: This is fine, important part is validation.

**Comment by SpaceManiac** [comment]
I notice a lot of `mob/user`; I would highly encourage you to tie buildmode to client rather than mob, to make things saner in the case of starting buildmode and then aghosting. All the `input()`s should still work (though a few use implicit `usr` rather than the `user` parameter).

**Comment by Crazylemon64** [comment]
@SpaceManiac As a `click_intercept`, it'll get nil'd out during mob login. \[[source](https://github.com/tgstation/tgstation/blob/master/code/modules/mob/login.dm#L44)\]

I was unable to entirely expunge usage of `usr` - for example, inside the `Click` proc, I am not aware of any alternate method of finding out the user who clicked the object. Buildmode already stored the holder as a client, and not a mob, so I did not need to change that. Without the full `click_intercept` infrastructure, however, the datum will be left dangling as a result of the login nil'ing.

**Comment by AnturK** [comment]
That click intercept removal on login seems completely unnecessary to me, it might be some remnant from before it was split into mob and client intercepts.

**Comment by Crazylemon64** [comment]
@AnturK Would you prefer that, or using redirect components to handle signals fired on mob login and logout?

**Comment by Crazylemon64** [comment]
Hmm. Other things that use `click_intercept`:
\* Certain spells
\* Manual turret control

How adversely affected would these be by removing the `click_intercept` nulling?

**Comment by AnturK** [comment]
These should be using **mob** click_intercept ? The one that shouldn't reset on Login is client one.

**Comment by Crazylemon64** [comment]
I don't see any discernible difference between these click intercepts - as far as I can tell, only one form of click intercept exists. Do I need to implement a system for both mob click intercepts and client click intercepts? How should I determine precedence if both are active?

**Review by AnturK — COMMENTED** [review]
How's this going ? Conflicts and you still didn't touch the client click intercept reset.

**Pull request closed without merging**